src/toplevel.pl

```diff
@@ -132,190 +136,501 @@
-    ).
 
+list_last([L0|Ls], L) :-
+    list_last(Ls, L0, L).
 
-submit_query_and_print_results_(Term, VarList) :-
-    '$get_b_value'(B),
-    call(Term),
-    write_eqs_and_read_input(B, VarList),
-    !.
-submit_query_and_print_results_(_, _) :-
-    %  clear attribute goal lists, which may be populated by
-    %  copy_term/3 prior to failure.
-    '$clear_attribute_goals',
-    write('false.'),
-    nl.
-
-
-submit_query_and_print_results(Term0, VarList) :-
-    expand_goal(call(Term0), user, Term),
-    !,
-    setup_call_cleanup(bb_put('$first_answer', true),
-                       submit_query_and_print_results_(Term, VarList),
-                       bb_put('$first_answer', false)).
-
-
-needs_bracketing(Value, Op) :-
-    catch((functor(Value, F, _),
-	       current_op(EqPrec, EqSpec, Op),
-	       current_op(FPrec, _, F)),
-	      _,
-	      false),
-    (  EqPrec < FPrec ->
-       true
-    ;  FPrec > 0, F == Value, graphic_token_char(F) ->
-       true
-    ;  F \== '.', '$quoted_token'(F) ->
-       true
-    ;  EqPrec == FPrec,
-       memberchk(EqSpec, [fx,xfx,yfx])
+list_last([], L, L).
+list_last([L0|Ls], _, L) :-
+    list_last(Ls, L0, L).
+
+eq_member(X, [Y|Ls]) :-
+    (   Ls == [] ->
+        X == Y
+    ;   X == Y
+    ;   eq_member(X, Ls)
     ).
 
-write_goal(G, VarList, MaxDepth) :-
-    (  G = (Var = Value) ->
-       (  var(Value) ->
-	      select((Var = _), VarList, NewVarList)
-       ;  VarList = NewVarList
-       ),
-       write(Var),
-       write(' = '),
-       (  needs_bracketing(Value, (=)) ->
-	      write('('),
-	      write_term(Value, [quoted(true), variable_names(NewVarList), max_depth(MaxDepth)]),
-	      write(')')
-       ;  write_term(Value, [quoted(true), variable_names(NewVarList), max_depth(MaxDepth)])
-       )
-    ;  G == [] ->
-       write('true')
-    ;  write_term(G, [quoted(true), variable_names(VarList), max_depth(MaxDepth)])
+chars(V) :- var(V), !, false.
+chars([]).
+chars([C|Cs]) :- atom(C), atom_length(C, 1), chars(Cs).
+
+spec_arity(fx, 1).
+spec_arity(fy, 1).
+spec_arity(xf, 1).
+spec_arity(yf, 1).
+spec_arity(xfx, 2).
+spec_arity(xfy, 2).
+spec_arity(yfx, 2).
+
+% variable_name([N = V|VNs], Name = Var) :-
+%     (   Var == V ->
+%         Name = N
+%     ;   variable_name(VNs, Name = Var)
+%     ).
+
+symbolic_control('\a', "\\a").
+symbolic_control('\b', "\\b").
+symbolic_control('\f', "\\f").
+symbolic_control('\n', "\\n").
+symbolic_control('\r', "\\f").
+symbolic_control('\t', "\\t").
+symbolic_control('\v', "\\v").
+% symbolic_control('\x0\', "\x0\"). % '.
+
+print_double_quoted([]) --> [].
+print_double_quoted([C|Cs]) -->
+    (   { symbolic_control(C, Cs0) } ->
+        Cs0
+    ;   { write_term_to_chars(C, [], Cs0) },
+        Cs0
+    ),
+    print_double_quoted(Cs).
+
+print_quoted([]) --> [].
+print_quoted([C|Cs]) -->
+    (   { C = '''' } ->
+        "''"
+    ;   { symbolic_control(C, Cs0) } ->
+        Cs0
+    ;   { write_term_to_chars(C, [], Cs0) },
```

Comment:
This is `format_("~w", [C])`.

Reply:
I wasn't sure if top-level should depend on `format`.

Reply:
It's been included now.